src/backend/distributed/master/master_modify_multiple_shards.c

```diff
@@ -104,18 +108,40 @@ master_modify_multiple_shards(PG_FUNCTION_ARGS)
 		relationId = RangeVarGetRelid(updateStatement->relation, NoLock, failOK);
 		EnsureTablePermissions(relationId, ACL_UPDATE);
 	}
+	else if (IsA(queryTreeNode, TruncateStmt))
+	{
+		TruncateStmt *truncateStatement = (TruncateStmt *) queryTreeNode;
+		List *relationList = truncateStatement->relations;
+		RangeVar *rangeVar = NULL;
+
+		if (list_length(relationList) != 1)
+		{
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("master_modify_multiple_shards() can truncate only "
+							"one table")));
+		}
+
+		rangeVar = (RangeVar *) linitial(relationList);
+		relationId = RangeVarGetRelid(rangeVar, NoLock, failOK);
+		EnsureTablePermissions(relationId, ACL_TRUNCATE);
+		validateModifyQuery = false;
+	}
 	else
 	{
-		ereport(ERROR, (errmsg("query \"%s\" is not a delete nor update statement",
-							   queryString)));
+		ereport(ERROR, (errmsg("query \"%s\" is not a delete, update, or truncate "
+							   "statement", queryString)));
 	}
 
 	CheckDistributedTable(relationId);
 
 	queryTreeList = pg_analyze_and_rewrite(queryTreeNode, queryString, NULL, 0);
 	modifyQuery = (Query *) linitial(queryTreeList);
 
-	ErrorIfModifyQueryNotSupported(modifyQuery);
+	if (validateModifyQuery)
```

Comment:
This boolean flag just adds confusion; instead, can we make this conditional `if (query->commandType != CMD_UTILITY)`?